File: demoGG/app/src/main/java/com/example/demogg/Download/DownloadData.java
package com.example.demogg.Download;

import android.net.Uri;
import android.os.AsyncTask;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DownloadData extends AsyncTask<String,Void,String>{
    Download download;
    List<HashMap<String,String>> listData;
    String link;
    boolean kiemtra = true;


    public DownloadData(Download download, List<HashMap<String, String>> listData, String link) {
        this.download = download;
        this.listData = listData;
        this.link = link;
        kiemtra = false;
    }

    public DownloadData(Download download, String link) {
        this.download = download;
        this.link = link;
        kiemtra = true;
    }

    @Override
    protected String doInBackground(String... strings) {
        String dulieu = "";
        if(!kiemtra){
            dulieu = methodPost(link);
        }else {
            dulieu = methodGet(link);
        }
        return dulieu;
    }

    private String methodPost(String link) {
        StringBuilder stringBuilder = new StringBuilder();
        String key = "";
        String value = "";
        String data = "";

        try {
            URL url = new URL(this.link);
            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
            httpURLConnection.setRequestMethod("POST");
            httpURLConnection.setDoOutput(true);
            httpURLConnection.setDoInput(true);

            Uri.Builder builder = new Uri.Builder();
            int count = listData.size();
            for (int i = 0; i < count;i++){
                for (Map.Entry<String,String> values : listData.get(i).entrySet()){
                    key = values.getKey();
                    value = values.getValue();
                }
                builder.appendQueryParameter(key,value);
            }
            String query = builder.build().getEncodedQuery();

            OutputStream outputStream = httpURLConnection.getOutputStream();
            OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputStream);
            BufferedWriter writer = new BufferedWriter(outputStreamWriter);
            writer.write(query);
            writer.flush();
            writer.close();
            outputStream.close();

            InputStream inputStream = httpURLConnection.getInputStream();
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
            String line = "";

            while ((line = bufferedReader.readLine()) != null){
                stringBuilder.append(line);
            }

            data = stringBuilder.toString();

        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return data;
    }

    private String methodGet(String link) {
        String data = "";

        try {
            URL url = new URL(this.link);
            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
            httpURLConnection.connect();

            InputStream inputStream = httpURLConnection.getInputStream();
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
            BufferedReader bufferedReader = new BufferedReader(inputStreamReader);

            StringBuilder stringBuilder = new StringBuilder();
            String line = "";

            while ((line = bufferedReader.readLine()) != null){
                stringBuilder.append(line + "\n");
            }

            data = stringBuilder.toString();
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data;
    }

    @Override
    protected void onPostExecute(String s) {
        super.onPostExecute(s);
        download.Download(s);
    }
}

File: LoiCuaBac2/app/src/main/java/com/example/loicuabac2/service/download/DownloadData.kt
package com.example.loicuabac2.service.download

import android.os.AsyncTask
import android.util.Log
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL


class DownloadData (private var downloadInterface : DownloadInterface, private var link : String):
    AsyncTask<String, Void, String>() {

    /*override fun doInBackground(vararg p0: String?): String {
        val url  = URL(link)
        val httpURLConnection = url.openConnection()
        httpURLConnection.connect()

        val inputStream = httpURLConnection.getInputStream()
        val inputStreamReader = InputStreamReader(inputStream)
        val bufferedReader = BufferedReader(inputStreamReader)

        val stringBuilder : StringBuilder = StringBuilder()
        val line = bufferedReader.readLine()

        while (line != null) {
            stringBuilder.append(line + "\n")
        }

        return stringBuilder.toString()
    }*/

    override fun onPostExecute(result: String?) {
        super.onPostExecute(result)
        if (result != null) {
            downloadInterface.getStringFormUrl(result)
        }

    }

    override fun doInBackground(vararg strings: String): String {
        var data = ""

        try {
            val url = URL(link)
            val httpURLConnection = url.openConnection() as HttpURLConnection
            httpURLConnection.connect()

            val inputStream : InputStream = httpURLConnection.inputStream
            val inputStreamReader: InputStreamReader = InputStreamReader(inputStream)
            val bufferedReader : BufferedReader = BufferedReader(inputStreamReader)

            val stringBuilder = StringBuilder()
            val line =  bufferedReader.readLine()

            if (line != null) {
                stringBuilder.append(line + "\n")
            }

            data = stringBuilder.toString()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return data
    }
}